src/main/scala/is/hail/annotations/aggregators/TransformedRegionValueAggregator.scala

```diff
@@ -0,0 +1,74 @@
+package is.hail.annotations.aggregators
+
+import is.hail.asm4s._
+import is.hail.expr._
+import is.hail.expr.ir._
+import is.hail.annotations._
+
+object TransformedRegionValueAggregator {
+  /**
+    * {@code t} is the type of value to which
+    * TransformedRegionValueAggregator.seqOp will be applied
+    *
+    * The argument to {@code makeTransform} is an IR that evalutes to the
+    * aggregable carrier struct. The output is an IR that evaluates to the
+    * desired input to {@code next.seqOp}.
+    *
+    **/
+  def apply(t: Type,
+    makeTransform: IR => IR,
+    next: RegionValueAggregator): TransformedRegionValueAggregator = {
+
+    val transform = makeTransform(In(0, t))
+    val out = makeCarrier(transform)
+    val fb = FunctionBuilder.functionBuilder[MemoryBuffer, Long, Boolean, Long]
+    Compile(out, fb)
+    new TransformedRegionValueAggregator(fb.result(), transform.typ, next)
+  }
+
+  private def carryingField = "x"
+
+  private def makeCarrier(element: IR): IR =
+    MakeStruct(Array((carryingField, element.typ, element)))
+}
+
+/**
+  * {@code getTransformer} is a thunk of a function that transforms an
+  * aggregation carrier struct to the desired input to {@code next.seqOp}.
+  *
+  **/
+class TransformedRegionValueAggregator(
+  getTransformer: () => AsmFunction3[MemoryBuffer, Long, Boolean, Long],
+  carriedType: Type,
```

Comment:
I don't think this should have a `Type`.  See my comment below.

I'm quite concerned about the code that will be generated here.  Can't you flatten the aggregator steps like you do with container types?  That is, `xs.map(x => f(x)).filter(y => p(y)).sum()` should generate something like:

```
  def seqOp(x: ...) { // region value type
    val y = f(x)
    if (p)
      acc += y
```

I'd very much like to avoid passing aggregated values through a bunch of objects like we do now in many places with `Iterator`s.

Reply:
I did not implement AggFilter or AggFlatmap in this PR, but here's my plan to do so. (Maybe this PR should require AggFilter and AggFlatmap? What is here now already seems like a big change)

So the normal form of an aggregator expression, AFAICT, is this:

    xs.map(e1)
      .filter(p1)
      .map(e2)
      .filter(p2)
      ...
      .flatMap(en)
      .agg()


I don't think we can further collapse these because we don't want to run `e2` on
elements that `filter(p1)` removes.

Given the above form, the ideal (psuedo) code does seems to be:

    x1 = e1(x)
    if (p1(x1))
      x2 = e2(x1)
      if (p2(x2))
        ...
        if (pn-1(xn-1))
          xn-1 = en-1(xn-2)
          rvArrayForeach(en(xn-1), agg.seqOp)

---

When I add AggFilter and AggFlatmap I need to handle three situations, AFAICT:

n map and n filters (`mapFilterPairs` must be non-empty):

    mapFilterPairs: Array[(IR => IR, IR => IR)]

n maps and n-1 filters:

    mapFilterPairs: Array[(IR => IR, IR => IR)]
    map: IR => IR

n maps, n filters and one flatMap step:

    mapFilterPairs: Array[(IR => IR, IR => IR)]
    flatMap: IR => IR

 - The map IRs return offsets

 - The filter IRs return Boolean

 - The flatMap returns an offset to a Array

 - The final map step must return an offset to a value of the type expected by seqOp

 - The flatMap step must return an offset to an array of values of the type expected by seqOp

---

I will implement these three cases as you suggest. I can stage the array iteration since it is known at aggregation-compile-time. If there is no flat-map the resulting bytecode should look more or less like this:

    x1 := e1(input2)
    p1(x1)
    IFEQ lend
    x2 := e2(x1)
    IFEQ lend
    ...
    xn
    input1
    INVOKEVIRTUAL Lis/hail/expr/RegionValueAggregator seqOp
    lend

input1 is the aggregator, input2 is the offset to the next element in the aggregatee.

Reply:
> I did not implement AggFilter or AggFlatmap in this PR

Sorry, the overall plan isn't clear from these chopped up PRs out of context.

What is `TransformedRegionValueAggregator` for?  To handle aggregator type conversions?  I don't think it should exist, then.  A type conversion is just a map: `gs.map(g => G.DP).stats()` should map to `gs.map(g => g.DP).map(dp => dp.toDouble).stats()`.  Like I requested for cast, these need to be put into the expr language explicitly so they can be optimized.

 > So the normal form of an aggregator expression, AFAICT, is this:

How do you always get the flatMap to the bottom?  It should certainly be possible to filter the output of a flat map.  If there is a general solution that is composable, it is better to use that than to require a specific form of the IR.  That will only created dependencies and difficulties upstream.  I would be inclined to support an arbitrary mixture of filter, map and flatMap.

Reply:
Aggregable map / flatMap may not need to exist at all -- the Hail2 interface will be something like `func.stats(m.GQ)`. Currently this translates into `gs.map(_ => g.GQ).stats()`, but maybe there's some better way to write the IR.

Reply:
Also, I should note that there is no flattening happening in the IR right now. That's totally something we should do, but right now maps and filters allocate new intermediate arrays. My  initial plan is to have a pass that fuses adjacent maps and adjacent filters. I don't have a plan to compile subsequent maps, filters and flatMaps together into a single-allocation line of code. I know we need to do that, but it's on my list after getting a functionality-complete IR (dicts, sets, linking to AST and friends, treating required types correctly). I'm happy to reprioritize if you think this stuff needs to happen first.

Reply:
>  However, map and flatMap are slightly more powerful because in @tpoterba current setup, filter goes at the beginning and flatMap goes at the end, am I remembering right?

The current hail2 interface has filter and explode as aggregable methods, so I think this would give you the same functionality. We'll still need explode, it just may be simpler to think about without lambdas.

Reply:
Ok `TransformedRegionValueAggregator` is now a lightweight class that lets me intercede on the seqOp of another aggregator.